### ISSUE TITLE
API for forwarding ngsim simulation

### ISSUE THREAD

**Issue by qianyi-sun** [huawei-noah/SMARTS#1407]
This is a request from @YuechengLiu, which is about adding an API in SMARTS like the following:

```
class SMARTS:
    def _reset(scenario, start_time=None):
        if start_time:
            self._elapsed_sim_time = start_time
        else:
            self._elapsed_sim_time = 0
```

**Comment by qianyi-sun** [comment]
@YuechengLiu, we've discussed this issue and please leave a comment to correct us if the following understanding is wrong: 

so this request is for allowing the simulation of ngsim data to forward to a specific time step, in order to skip some `sim_time` from the beginning, is this right?

**Comment by YuechengLiu** [comment]
> @YuechengLiu, we've discussed this issue and please leave a comment to correct us if the following understanding is wrong:
> 
> so this request is for allowing the simulation of ngsim data to forward to a specific time step, in order to skip some `sim_time` from the beginning, is this right?

Exactly. And we can discuss about this issue in detail later.

**Comment by YuechengLiu** [comment]
In the case that some of of vehicles are controlled by user and the others are loaded from traffic histories, (just like what has been done in: https://github.com/huawei-noah/SMARTS/blob/master/examples/history_vehicles_replacement_for_imitation_learning.py)
.

I find that the SMARTS._reset() are **not able to get the observations of ego vehicles immediately**, and SMARTS while keep calling SMARTS._step() until ego obeservations are returned. 
https://github.com/huawei-noah/SMARTS/blob/3ef5650b04ac6fb7145cf4e23d5534d73e0929fc/smarts/core/smarts.py#L434
This process will last for 2 steps ~ 20steps。During the procedure, the internal clock (SMARTS._elapsed_sim_time) is still keep going，and the vehicles (which are not controlled by user) provided by the traffic provider is loaded according to the new elapsed_sim_time. However, the state of ego vehicles are still old and **can not match** its initial state from the traffic history under the new internal clock.

**Comment by Gamenot** [comment]
## Understanding

I understand the issues to be the following:
- Simulation time increments during reset.
- Initially specified ego state becomes old during that increment (or are emitted at a location later where they should have captured a vehicle)
- Wasteful time spent simulating in reset.

So your request is to set the jump the simulation time at beginning of reset in an attempt to skip all steps before ego observations start.

## Issue

Then the problem is the following even if we get the `traffic_history_provider` to move to a different time:

```python
>>> ## assuming first agent mission start time is 0
>>> start_time = 10
>>> traffic_history_provider = smarts.get_provider_by_type(TrafficHistoryProvider)
>>> traffic_history_provider.start_time = start_time
>>> smarts_obs = smarts.reset(scenario) # <-- issue occurs here and does not return immediately
>>> thp_time = start_time + smarts.elapsed_sim_time
>>> expected_time = start_time + smarts.fixed_timestep_sec
>>> assert thp_time <= expected time
~AssertionError 10.5 <= 10.1
```

## Question

Where are the missions being specified from:
- Are the ego missions being specified through scenario studio?
- Are they dynamically generated as with https://github.com/huawei-noah/SMARTS/blob/master/examples/dynamic_history_vehicles_replacement.py?

## Interface

```python
## skip the simulation clock ahead to acquire the observations immediately
smarts.reset(scenario, start_time= start_time) # traffic_history_provider ends up at start_time
```

It is also my understanding that what you might be able to work with the following interface:

```python
traffic_history_provider.start_time = <time>
## All simulation until the time that the ego observations become available
smarts.reset(scenario, skip_simulation = True) # first ego observes immediately so traffic_history_provider ends up at <time>
```

**Comment by YuechengLiu** [comment]
@Gamenot  Thank for your help

### Are the ego missions specified through scenario studio?
Here is my ego_missions:

```python
  def _vehicle_pos_between(self, vehicle_id: str, start_time: float, end_time: float):
      """Find the vehicle states between the given history times."""
      query = """SELECT T.position_x, T.position_y, T.sim_time
                 FROM Trajectory AS T
                 WHERE T.vehicle_id = ? AND T.sim_time > ? AND T.sim_time <= ? 
                 ORDER BY T.sim_time DESC"""
      rows = self.scenario._traffic_history._query_list(
          query, (vehicle_id, start_time, end_time)
      )
      return (VehiclePosition(*row) for row in rows)

  def _discover_sliced_vehicle_missions(self):
      """Retrieves the missions of traffic history vehicles."""
      vehicle_missions = {}
      for vehicle_id in self.vehicle_ids:

          # the start_time is predefined by myself according to some rule (not the first seen time of the vehicle).

          start_time = self.vehicle_start_times[vehicle_id]
          rows = self._vehicle_pos_between(
              vehicle_id,
              start_time=start_time - self.smarts._last_dt,
              end_time=start_time,
          )
          rows = list(rows)
          assert len(rows) > 0, rows
          sim_time = rows[0].sim_time
          start, speed = self.scenario.get_vehicle_start_at_time(vehicle_id, sim_time)
          entry_tactic = default_entry_tactic(speed)
          veh_config_type = self.scenario._traffic_history.vehicle_config_type(
              vehicle_id
          )
          veh_dims = self.scenario._traffic_history.vehicle_dims(vehicle_id)
          vehicle_missions[vehicle_id] = Mission(
              start=start,
              entry_tactic=entry_tactic,
              goal=TraverseGoal(self.scenario.road_map),
              start_time=sim_time,
              vehicle_spec=VehicleSpec(
                  veh_id=vehicle_id,
                  veh_config_type=veh_config_type,
                  dimensions=veh_dims,
              ),
          )

```
And, 
```python
self.history_start_time = np.inf
# here we should not that self.vehicle_start_times are predefined by myself. (not the first-seen-times of the vehicles)
for agent_id in self.agent_ids:
    vehicle_id = self.aid_to_vid[agent_id]
    agent_interfaces[agent_id] = self.agent_spec.interface
    self.history_start_time = min(
        self.history_start_time, self.vehicle_start_times[vehicle_id]
    )

self.vehicle_missions = self._discover_sliced_vehicle_missions()
self.ego_missions = {}

# the start_time of the mission of the first-start-vehicle  is set to zero.
for agent_id in self.agent_ids:
    vehicle_id = self.aid_to_vid[agent_id]
    self.ego_missions[agent_id] = replace(
        self.vehicle_missions[vehicle_id],
        start_time=self.vehicle_start_times[vehicle_id]
        - self.history_start_time,
    )

self.scenario.set_ego_missions(self.ego_missions)
self.smarts.switch_ego_agents(agent_interfaces)
raw_observation_n = self.smarts.reset(
    self.scenario, start_time=self.history_start_time
)  # set  SMARTS._elapsed_sim_time to be the earliest start_time (predefined by myself) of controlled vehicles.
```
### Do the missions start relative to 0 and move the traffic_history_provider starting time to ?
I set the mission of the earliest ego vehicle to zero, but I **do not move** the the traffic_history_provider starting time (since I directly change the smarts._elapsed_sim_time and thus they are synchronized at the beginning, before the smarts' keep calling of SMARTS._step() to get the first ego observation)

**Comment by Gamenot** [comment]
@YuechengLiu Thank you for clarifying the issue. It appears that the default entry tactic is failing for relying on time delta rather than the current simulation time.

## Action

Your request seems reasonable. I will add the interface and fix the issues it causes with parts of SMARTS that collect time deltas to stay synchronized.

## Consideration
If I find myself with a little more time I will add an interface for acquiring a vehicle using `vehicle_id`+`time`.

```python
entry_tactic = VehicleIDEntryTactic(
    vehicle_id, ## vehicle id
    start_time, ## acquire this vehicle if it is at or after this time
    last_time, ## give up on aquiring this vehicle as an ego if it is after this time
)
```

The idea in this case would be for smarts to directly reserve that vehicle as an ego vehicle and take acquire that vehicle for the agent if the vehicle is simulated in that time range (skipping that time would also skip acquiring it.)

**Comment by YuechengLiu** [comment]
@Gamenot  Thanks for the quick reply, and looking forward to the solution. Thanks again!

**Comment by Gamenot** [comment]
Issue has been addressed with #1412; continued work in follow-up issue #1416.
